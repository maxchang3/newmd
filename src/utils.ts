import yaml from 'js-yaml'
import TOML from 'smol-toml'

export const getFrontmatter = (object: Record<string, unknown>, toml: boolean) => {
    const fence = toml ? '+++' : '---'
    const frontmatter = toml ? TOML.stringify(object) : yaml.dump(object).trim()
    return `${fence}\n${frontmatter}\n${fence}\n`
}
