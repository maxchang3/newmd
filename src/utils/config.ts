import type { Config } from '@/types'
import { DEFAULT_CONFIG } from '@/consts'
import { deepmerge } from 'deepmerge-ts'
import { createConfigLoader } from 'unconfig'

export const resolveConfig = async (config?: Config, options?: {
    cwd?: string
}) => {
    const loader = createConfigLoader<Config>({
        sources: [
            {
                files: 'newmd.config',
                extensions: ['ts', 'mjs', 'js'],
            },
        ],
        cwd: options?.cwd,
    })

    const loadResult = await loader.load()

    // Respect config file when no explicit options are provided
    if (!loadResult.sources.length) return deepmerge(DEFAULT_CONFIG, config) as Required<Config>

    return deepmerge(deepmerge(DEFAULT_CONFIG, loadResult.config), config) as Required<Config>
}
