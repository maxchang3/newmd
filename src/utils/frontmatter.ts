import type { StringKeys, ZodStringKeys } from '@/types'
import type { AnyZodObject } from 'zod'
import yaml from 'js-yaml'
import toml from 'smol-toml'
import { init as createSchemaDefaults } from 'zod-empty'

export type FrontmatterType = 'yaml' | 'toml'

interface FrontmatterConfig<Metadata extends Record<string, any>> {
    /**
     * Base metadata for the markdown file, will be used to generate the frontmatter.
     */
    baseMetadata: Metadata
    /**
     * Key in the metadata object that will be used as the title, the value must be a string.
     */
    titleKey: StringKeys<Metadata>
    /**
     * Type of frontmatter to generate.
     *
     * @defaultValue "yaml"
     */
    type?: FrontmatterType
}

export class Frontmatter<Metadata extends Record<string, any>> {
    config: FrontmatterConfig<Metadata>

    constructor({ baseMetadata, titleKey, type = 'yaml' }: typeof this.config) {
        this.config = { baseMetadata, type, titleKey }
    }

    static fromZodSchema<Schema extends AnyZodObject>(schema: Schema, options: {
        titleKey: ZodStringKeys<Schema>
        type?: FrontmatterType
    }) {
        const { titleKey, type = 'yaml' } = options
        const defaultData = createSchemaDefaults(schema)
        return new Frontmatter({ baseMetadata: defaultData, type, titleKey })
    }

    toString(title: string) {
        const { titleKey, baseMetadata } = this.config
        const metadata = structuredClone(baseMetadata)
        const filteredMetadata = Object.fromEntries(
            Object.entries(metadata).filter(([_, value]) => value !== null),
        )
        if (titleKey in filteredMetadata) {
            // @ts-expect-error titleKey is a valid key in metadata
            filteredMetadata[titleKey] = title
        } else {
            throw new Error(`Title key "${String(titleKey)}" not found in metadata`)
        }
        const isTOML = this.config.type === 'toml'
        const fence = isTOML ? '+++' : '---'
        const frontmatter = isTOML
            ? toml.stringify(filteredMetadata)
            : yaml.dump(filteredMetadata).trim()
        return `${fence}\n${frontmatter}\n${fence}\n`
    }
}
