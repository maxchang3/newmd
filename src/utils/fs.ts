import fs from 'node:fs/promises'
import { resolve } from 'pathe'

export const writeMarkdownFile = async (options: {
    filename: string
    content: string
    path: string
    cwd?: string
}) => {
    const { filename, content, path, cwd } = options

    const outputDir = resolve(cwd ?? '', path)
    const filepath = resolve(outputDir, `${filename}.md`)

    await fs.mkdir(outputDir, { recursive: true })
    await fs.writeFile(filepath, content)

    return filepath
}
