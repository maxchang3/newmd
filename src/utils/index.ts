export * from './frontmatter'
export * from './fs'
