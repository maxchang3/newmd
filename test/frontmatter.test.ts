import { describe, expect, it } from 'vitest'
import { z } from 'zod'
import { Frontmatter } from '@/utils'

describe('frontmatter', () => {
    it('should generate frontmatter in yaml format', () => {
        const frontmatter = new Frontmatter({
            baseMetadata: { title: '' },
            titleKey: 'title',
        })
        const result = frontmatter.toString('Hello World')
        expect(result).toBe('---\ntitle: Hello World\n---\n')
    })

    it('should generate frontmatter with custom title key', () => {
        const frontmatter = new Frontmatter({
            baseMetadata: { headline: '' },
            titleKey: 'headline',
        })
        const result = frontmatter.toString('Hello World')
        expect(result).toBe('---\nheadline: Hello World\n---\n')
    })

    it('should generate frontmatter in toml format', () => {
        const frontmatter = new Frontmatter({
            baseMetadata: { title: '' },
            titleKey: 'title',
            type: 'toml',
        })
        const result = frontmatter.toString('Hello World')
        expect(result).toBe('+++\ntitle = "Hello World"\n+++\n')
    })

    it('should throw error if title key is not found in metadata', () => {
        const frontmatter = new Frontmatter({
            baseMetadata: {},
            // @ts-expect-error title is not a valid key in metadata
            titleKey: 'title',
        })
        expect(() => frontmatter.toString('Hello World')).toThrowErrorMatchingSnapshot()
    })

    it('should generate frontmatter from zod schema', () => {
        const schema = z.object({
            title: z.string(),
            description: z.string().optional(),
        })
        const frontmatter = Frontmatter.fromZodSchema(schema, {
            titleKey: 'title',
        })
        const result = frontmatter.toString('Hello World')
        expect(result).toBe("---\ntitle: Hello World\ndescription: ''\n---\n")
    })
})
