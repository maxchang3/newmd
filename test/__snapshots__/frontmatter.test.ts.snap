// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`frontmatter > should throw error if title key is not found in metadata 1`] = `[Error: Title key "title" not found in metadata]`;
